File: app/src/main/java/com/example/taxconnect/services/AgoraManager.java
package com.example.taxconnect.services;

import android.content.Context;
import android.util.Log;
import io.agora.rtc2.Constants;
import io.agora.rtc2.IRtcEngineEventHandler;
import io.agora.rtc2.RtcEngine;

public class AgoraManager {
    private static final String TAG = "AgoraManager";
    private static final String APP_ID = com.example.taxconnect.BuildConfig.AGORA_APP_ID;
    private static final String APP_CERTIFICATE = null; // Set to null for App ID only mode
    
    private static AgoraManager instance;
    private RtcEngine rtcEngine;
    private final Context context;

    private AgoraManager(Context context) {
        this.context = context.getApplicationContext();
        initializeAgoraEngine();
    }

    public static synchronized AgoraManager getInstance(Context context) {
        if (instance == null) {
            instance = new AgoraManager(context);
        }
        return instance;
    }

    private void initializeAgoraEngine() {
        try {
            rtcEngine = RtcEngine.create(context, APP_ID, new IRtcEngineEventHandler() {
                @Override
                public void onJoinChannelSuccess(String channel, int uid, int elapsed) {
                    Log.d(TAG, "onJoinChannelSuccess: " + channel + " uid: " + uid);
                }

                @Override
                public void onUserJoined(int uid, int elapsed) {
                    Log.d(TAG, "onUserJoined: " + uid);
                }

                @Override
                public void onUserOffline(int uid, int reason) {
                    Log.d(TAG, "onUserOffline: " + uid);
                }
            });
            // Enable video by default for video calls
            rtcEngine.enableVideo();
        } catch (Exception e) {
            Log.e(TAG, "Error initializing Agora RtcEngine", e);
        }
    }

    public void joinChannel(String channelName, String token, int uid) {
        if (rtcEngine != null) {
            rtcEngine.joinChannel(token, channelName, "Extra Optional Data", uid);
        }
    }

    public void leaveChannel() {
        if (rtcEngine != null) {
            rtcEngine.leaveChannel();
        }
    }
    
    public RtcEngine getRtcEngine() {
        return rtcEngine;
    }

    public void destroy() {
        if (rtcEngine != null) {
            RtcEngine.destroy();
            rtcEngine = null;
        }
        instance = null;
    }
}

File: tax/TAX-connect-prototype/tax/app/src/main/java/com/example/taxconnect/data/services/AgoraManager.kt
package com.example.taxconnect.data.services

import android.content.Context
import android.util.Log
import com.example.taxconnect.BuildConfig
import io.agora.rtc2.IRtcEngineEventHandler
import io.agora.rtc2.RtcEngine

class AgoraManager private constructor(context: Context) {
    private val appContext: Context = context.applicationContext
    var rtcEngine: RtcEngine? = null
        private set

    init {
        initializeAgoraEngine()
    }

    private fun initializeAgoraEngine() {
        try {
            rtcEngine = RtcEngine.create(appContext, APP_ID, object : IRtcEngineEventHandler() {
                override fun onJoinChannelSuccess(channel: String, uid: Int, elapsed: Int) {
                    Log.d(TAG, "onJoinChannelSuccess: $channel uid: $uid")
                }

                override fun onUserJoined(uid: Int, elapsed: Int) {
                    Log.d(TAG, "onUserJoined: $uid")
                }

                override fun onUserOffline(uid: Int, reason: Int) {
                    Log.d(TAG, "onUserOffline: $uid")
                }
            })
            // Enable video by default for video calls
            rtcEngine?.enableVideo()
        } catch (e: Exception) {
            Log.e(TAG, "Error initializing Agora RtcEngine", e)
        }
    }

    fun joinChannel(channelName: String?, token: String?, uid: Int) {
        rtcEngine?.joinChannel(token, channelName, "Extra Optional Data", uid)
    }

    fun leaveChannel() {
        rtcEngine?.leaveChannel()
    }

    fun destroy() {
        if (rtcEngine != null) {
            RtcEngine.destroy()
            rtcEngine = null
        }
        instance = null
    }

    companion object {
        private const val TAG = "AgoraManager"
        private val APP_ID: String = BuildConfig.AGORA_APP_ID
        
        @Volatile
        private var instance: AgoraManager? = null

        @JvmStatic
        fun getInstance(context: Context): AgoraManager {
            return instance ?: synchronized(this) {
                instance ?: AgoraManager(context).also { instance = it }
            }
        }
    }
}
